### DISCUSSION TITLE
Golang dependencies on generated code

### DISCUSSION THREAD

**Discussion by ronnylt** [renovatebot/renovate#23471] [Suggest an Idea]
**What Renovate type are you using?**
The hosted App

**Describe the bug**
When having a Golang codebase with dependencies on generated code (i.e. Protocol Buffers) a previous step is required to generate the code before running `go mod ...` . If the generated code is not present, the `go mod ...` will fail.

**To Reproduce**
Any Golang codebase with dependencies on generated code, for example a `go generate` invocation required before running `go mod ...`

**Expected behavior**
Modules updated.

**Screenshots**

```

Command failed: docker run --rm -v /mnt/renovate/gh/xxxx/xxxxx:/mnt/renovate/gh/xxxx/xxxx -v /tmp/renovate-cache/others/go:/tmp/renovate-cache/others/go -e GOPATH -e GOPROXY -e CGO_ENABLED=0 -w /mnt/renovate/gh/xxx/xxx renovate/go bash -c "git config --global url.\"https://x-access-token:**redacted**@github.com/\".insteadOf \"https://github.com/\" && go get ./..."
build xxx/cmd/xxx cannot load github.com/xxx/protos/go/xxxx: no matching versions for query "latest"


```

**Comment by rarkins** [comment]
Will `go get` still succeed? Or it will also fail?

**Comment by diagonalfish** [comment]
I can confirm this problem in go 1.13 - 'go get' or any other 'go' command that acts on the module will fail due to the missing generated code with the error in the OP.

I was considering submitting a feature request for "preUpdateActions" that'd be done before attempting to do updates, which would allow you to work around sort of problem.

**Comment by xens** [comment]
I'm wondering if there's any progress on that one ? I'd love to help with a PR but I've never contributed to Renovate so far, so any tips / guidance is welcome.

**Comment by rarkins** [comment]
This is still in the "Needs Requirements" category, meaning we need to work out how we'd do it. We already have the concept of [postUpdateTasks](https://docs.renovatebot.com/configuration-options/#postupgradetasks) so the suggestion of preUpdateTasks seems logical. This would be applicable for *self hosted* only so that you have full flexibility of what you need to run.

**Comment by xens** [comment]
@rarkins so if I understand correctly the concept of preUpdateTasks would be a kind of workaround as it'd allows us to pre-execute the missing `go generate` command before Renovate issue the `go mod` one. Do we also want to work on fixing the logic inside the golang manager ?

**Comment by rarkins** [comment]
No, I think we have a misunderstanding. At this point it's *intended* that we don't allow arbitrary code execution by default, such as `go generate` would allow. I'm suggesting that you can *add a feature* to allow arbitrary code execution prior to updating in case you are running self-hosted and don't mind the risk.

By the way, the original description in this issue seems a little wrong because it says `go generate` is required before any `go mod` however as per the log we are actually running `go get` and not any `go mod` command. `go mod` is only run if people opt into go mod tidy.

**Comment by xens** [comment]
Ok gotcha thanks for the explanation that makes a lot of sense.

**Comment by rarkins** [comment]
Right now we have:
- `postUpdateOptions`: simple, safe opt-in things to be run after updating, if opted into. e.g. `gomodTidy` means running `go mod tidy` after `go get`.
- `postUpgradeTasks`: similar to the above, but can be completely arbitrary code
- `allowedPostUpgradeCommands`: something which lets bot owners have a whitelist of allowed commands/patterns

We could perhaps rename and reuse `allowedPostUpgradeCommands` e.g. into `allowedUpgradeCommands` to be shared between `preUpdateTasks` and `postUpdateTasks`, as there doesn't seem to be any reason to differentiate between them from a security point of view.

**Comment by xens** [comment]
> postUpdateTasks: similar to the above, but can be completely arbitrary code

you mean `postUpgradeTasks` I guess, but yeah your proposal sounds good !

**Comment by rarkins** [comment]
Thanks, corrected

**Comment by ghost** [comment]
Any progress on this issue?

We have a problem with renovate and gererated modules. We have a module, which is referenced by an import in code but will generated on build time with make. So if renovate tries to go throw imports with `go get` it cannot resolve the import. There is any solution on that? Can I execute `make build` before renovate fetching dependencies? This problem should have many projects working with module generation.

I couldn't find any `preUpgradeTasks`

**Comment by HonkingGoose** [comment]
We want to have actionable Issues on our repository. We're still figuring out the requirements for this issue after four years. So I'm converting this to a Discussion.

Once we know and understand what to do, the maintainers will create a new issue.

**Comment by hfjn** [comment]
We are having the same issue in our code base and would require some way of preparing the code base so renovate can properly resolve all dependencies.

**Comment by clifg** [comment]
We're running into the same issue. For modules local to the repo, we can check in placeholders to satisfy the `go get` tooling but the generated code (swagger, other mockgen stuff) also then has imports needed and those aren't present when renovatebot does its work...

Edit: I wonder if a (hacky, but hopefully temporary) workaround would be to use placeholders for local generated modules like mocks and such, so the `go get` will succeed, and then use `postUpgradeTasks` to then run the full code generation and then re-do the `go get` command. Basically get renovate happy with the current repo w/o any generated code and then re-run the whole thing again as part of `postUpgradeTasks`. Typing it out, I really do hate it 😂

**Comment by philipparndt** [comment]
I've found another workaround:

You can use this configuration:
```json
{
  "goGetDirs": ["go@latest"],
  "postUpdateOptions": [
      "gomodTidy"
  ]
}
```

It will reduce the impact of `go get -d -t ./...` to only update the go version. If you like to manage your go version by hand, you can also add the version you like to use instead of `latest`.

I don't love it, as it is a hack, but it will do the job.

**Reply by noi** [reply]
Thanks, your ideas helped me. And I made a few changes as follows:

```
"goGetDirs": ["example.com/mod@none"]
```

This is about removing dependencies that don't exist.

```
$ go help get | grep "go get example.com/mod@none" -B 2
To remove a dependency on a module and downgrade modules that require it:

	go get example.com/mod@none
```

**Reply by d-rk** [reply]
Thanks, this also helped me. I now have the following packageRule:
```
		{
			"matchManagers": ["gomod"],
			"goGetDirs": ["mydomain.com/nonexistant-module@none"],
			"postUpgradeTasks": {
				"commands": [
					"cd {{packageFileDir}} && install-tool golang 1.24.6",
					"cd {{packageFileDir}} && go mod tidy -e",
					"cd {{packageFileDir}} && go generate",
					"cd {{packageFileDir}} && go mod tidy"
				]
			}
		},
```

- `"goGetDirs": ["mydomain.com/nonexistant-module@none"],` leads to a disabling of the `go get` command renovate itself is doing to update the `go.sum` file.
- Afterwards you can do basically anything you need to get the `go.sum` updated correctly.

In my case:
- I use `cd {{packageFileDir}}` because I have multiple `go.mod` files in the repo
- I use `go mod tidy -e` before `go generate` because I use `go tool` directive in go generate and renovate might update these tools.
- I run `go generate` to generate additional sources (mocks, openapi interfaces)
- I run `go mod tidy` again when the generated files are present.

I know its ugly, but it seems to work at least. 
Perhaps this helps someone.

**Comment by CynanX** [comment]
I've stumbled across this issue today - thanks for the tips on workarounds - however, any updates on an actual solution, like being able to specify pre commands that can then be used to generate the missing files?

**Reply by jamietanna** [reply]
Per https://github.com/renovatebot/renovate/discussions/40237, https://github.com/renovatebot/renovate/discussions/24893 would be the solution here, but we've not yet defined what this would look like

**Comment by jamietanna** [comment]
Other related issues:

- https://github.com/renovatebot/renovate/discussions/40834

**Comment by gabrpt** [comment]
Hello, we are also running into this issue. Any update on an actual solution ?

**Reply by jamietanna** [reply]
Short answer is "no" - this isn't something we've got on our immediate roadmap (as far as we have a roadmap that is)

Also note that even if this were implemented, we will not allow this for non-paying customers due to the inherent security risks that can occur from running `go generate` (and why we've gated `postUpdateOptions=goGenerate` behind [`allowedUnsafeExecutions`](https://docs.renovatebot.com/self-hosted-configuration/#allowedunsafeexecutions))

---

[My](https://www.jvt.me/posts/2022/05/05/commit-go-generate/) personal recommendation is to commit your generated code - this not only allows tools like Renovate to know what dependencies you have, as well as making it easier for others using your project, and i.e. allowing folks to see diffs in the commits that change generated code

Note that if you're not committing generated code (and therefore have dependencies that may be used by it) there will be many tools that can't understand this - for instance security vulnerability scanners

---

As a separate point, we generally keep our Discussions up-to-date when there are updates - if there's no recent discussion, it means there's no progress, so there's no need to nudge an old thread
